crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -511,13 +311,57 @@ impl QuiltVersionV1 {
     const QUILT_VERSION_BYTE: u8 = 0x01;
     const BLOB_HEADER_SIZE: usize = 6;
 
+    /// Decodes the quilt index from a column data source.
+    pub fn decode_quilt_index<T>(
+        data_source: &T,
+        column_size: usize,
+    ) -> Result<QuiltIndexV1, QuiltError>
+    where
+        T: QuiltColumnRangeReader,
+    {
+        let quilt_version_bytes =
+            data_source.range_read_from_columns(0, 0, QUILT_VERSION_BYTES_LENGTH)?;
+        assert!(quilt_version_bytes.len() == QUILT_VERSION_BYTES_LENGTH);
+        utils::check_quilt_version::<QuiltVersionV1>(&quilt_version_bytes)?;
+
+        let size_bytes = data_source.range_read_from_columns(
+            0,
+            QUILT_VERSION_BYTES_LENGTH,
+            QUILT_INDEX_SIZE_BYTES_LENGTH,
+        )?;
+        assert!(size_bytes.len() == QUILT_INDEX_SIZE_BYTES_LENGTH);
+
+        let index_size = usize::try_from(u32::from_le_bytes(
+            size_bytes
+                .try_into()
+                .map_err(|_| QuiltError::FailedToExtractQuiltIndexSize)?,
+        ))
+        .map_err(|_| QuiltError::FailedToExtractQuiltIndexSize)?;
+
+        let index_bytes =
+            data_source.range_read_from_columns(0, QUILT_INDEX_PREFIX_SIZE, index_size)?;
+        assert!(index_bytes.len() == index_size);
+
+        // Decode the QuiltIndexV1.
+        let mut quilt_index: QuiltIndexV1 = bcs::from_bytes(&index_bytes)
+            .map_err(|e| QuiltError::QuiltIndexSerDerError(e.to_string()))?;
```

Comment:
Nit: This particular error type could directly include the BCS error instead of its string representation. Also, the error message in the only other case is a bit confusing (`failed to decode the quilt index: failed to serialize quilt index: [...]`).